### PR TITLE
Use OnnxModels for private inference

### PR DESCRIPTION
**Pull request by gmuraru**
## Description
Define and use ```OnnxModels``` serialized models.
The parties know the model architecture and one of the parties has data that should be run through the model.
The local party does not have any idea about the architecture or the data.

The local party would only receive the result for the inference step.

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Added test to check the functionality

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pip-dep/requirements.txt [hunk 1]
```diff
@@ -12,7 +12,7 @@ psutil==5.7.0
 requests~=2.22.0
 RestrictedPython~=5.0
 scipy~=1.4.1
-syft-proto==0.4.7
+git+https://github.com/gmuraru/syft-proto.git@d8a4069457f65891ff02ad7e565dace8cd7b9a3f#egg=syft-proto
```
Comment (gmuraru) [inline pip-dep/requirements.txt:15]: This needs to be merged first.
```OnnxModel``` has protobuf + serde support.

File: test/serde/protobuf/test_protobuf_serde_full.py [hunk 2]
```diff
@@ -46,6 +46,10 @@
 samples[syft.generic.pointers.pointer_dataset.PointerDataset] = make_pointerdataset
 samples[syft.generic.string.String] = make_string
 
+# OnnxModel
+samples[syft.frameworks.crypten.model.OnnxModel] = make_onnxmodel
```
Comment (gmuraru) [inline test/serde/protobuf/test_protobuf_serde_full.py:50]: Test Protobuf

File: test/serde/msgpack/test_msgpack_serde_full.py [hunk 3]
```diff
@@ -69,6 +69,8 @@
 ] = make_fixedprecisiontensor
 samples[syft.frameworks.torch.tensors.interpreters.private.PrivateTensor] = make_privatetensor
 
+samples[syft.frameworks.crypten.model.OnnxModel] = make_onnxmodel
```
Comment (gmuraru) [inline test/serde/msgpack/test_msgpack_serde_full.py:72]: Test MsgPack

File: syft/frameworks/crypten/model.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    def __init__(
+        self,
+        serialized_model: bytes = None,
+        id: int = None,
+        owner: "syft.workers.AbstractWorker" = None,
+        tags: List[str] = None,
+        description: str = None,
+    ):
+        super(OnnxModel, self).__init__(id, owner, tags, description, None)
+        self.serialized_model = serialized_model
```
Comment (youben11) [inline syft/frameworks/crypten/model.py:27]: Shouldn't we also store the model's weights as discussed in #3776 ?

**Comment by kamathhrishi** [comment]
@gmuraru Great feature! Would love to see how this comes up.

File: pip-dep/requirements.txt (hunk 1 above)
Comment (LaRiffle) [reply pip-dep/requirements.txt:15]: Merged to syft-proto you can update the requirement to master :)

File: syft/frameworks/crypten/__init__.py [hunk 5]
```diff
@@ -70,3 +72,21 @@ def load(tag: str, src: int, **kwargs):
         result = crypten.load_from_party(preloaded=-1, src=src, **kwargs)
 
     return result
+
+
+def load_model(tag: str):
+    """
+    ATTENTION: All the workers that are part of the CrypTen computation
+    should have the model
+    This method should dissapear once CrypTen has support to load models
```
Comment (LaRiffle) [inline syft/frameworks/crypten/__init__.py:81]:
```suggestion
    This method should disappear once CrypTen has support to load models
```

File: syft/frameworks/crypten/__init__.py [hunk 6]
```diff
@@ -70,3 +72,21 @@ def load(tag: str, src: int, **kwargs):
         result = crypten.load_from_party(preloaded=-1, src=src, **kwargs)
 
     return result
+
+
+def load_model(tag: str):
+    """
+    ATTENTION: All the workers that are part of the CrypTen computation
```
Comment (LaRiffle) [inline syft/frameworks/crypten/__init__.py:79]:
```suggestion
    WARNING: All the workers that are part of the CrypTen computation
```

File: syft/frameworks/crypten/hook/hook.py [hunk 7]
```diff
@@ -7,6 +7,19 @@
 
 
 class CrypTenPlanBuild(object):
+    """
+    TODO: When the problem converting Onnx serialized models to PyTorch is solved
```
Comment (LaRiffle) [inline syft/frameworks/crypten/hook/hook.py:11]: Should this TODO info be at this place in the code?

File: syft/frameworks/crypten/hook/hook.py [hunk 8]
```diff
@@ -7,6 +7,19 @@
...
+    we can have serialized models on a SINGLE worker.
+
+    Those model will be known by a single party and when the party will call
+    the crypten.load function it will deserialize the model to a Torch one
+    and share it with the other parties.
+
+    In this scenario, the worker that started the computation will not know about the
+    instrinsics of the architecture and ONLY ONE worker can have knowledge about the
+    model
```
Comment (LaRiffle) [inline syft/frameworks/crypten/hook/hook.py:20]:
```suggestion
    This model will be known by a single party and when the party will call
    the crypten.load function it will deserialize the model to a PyTorch one
    and share it with the other parties.

    In this scenario, the worker that started the computation will not know about the
    instrinsics of the architecture and *only one* worker can have knowledge about the
    model
```

File: syft/frameworks/crypten/model.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+import torch
+
+import syft
+
+from syft.workers.abstract import AbstractWorker
+from syft.workers.base import BaseWorker
+from syft.frameworks.crypten import utils
+
+from syft.generic.pointers.object_pointer import ObjectPointer
+from syft.generic.abstract.sendable import AbstractSendable
```
Comment (LaRiffle) [inline syft/frameworks/crypten/model.py:12]:
```suggestion
from syft.frameworks.crypten import utils
from syft.generic.pointers.object_pointer import ObjectPointer
from syft.generic.abstract.sendable import AbstractSendable
from syft.workers.abstract import AbstractWorker
from syft.workers.base import BaseWorker
```

File: test/crypten/test_context.py [hunk 10]
```diff
@@ -219,14 +220,46 @@ def test_context_plan_with_model(workers):
...
+        model.decrypt()
         out = out.get_plain_text()
-        return model, out  # noqa: F821
+        return model, out
 
     result = plan_func_model()
     assert th.all(result[0][1] == result[1][1])
+
+
+def test_context_plan_with_model_private(workers):
```
Comment (LaRiffle) [inline test/crypten/test_context.py:236]: The model here compared to the previous test could be private (ie hidden) to the local worker?
If so you can add a line to explain the difference with the previous test!

File: syft/frameworks/crypten/model.py (hunk 4 above)
Comment (gmuraru) [reply syft/frameworks/crypten/model.py:27]: Based on our discussion, I will create another PR for doing that -- will link it to #3776 when doing that

File: syft/frameworks/crypten/hook/hook.py (hunk 7 above)
Comment (gmuraru) [reply syft/frameworks/crypten/hook/hook.py:11]: I think it is best to be in the ```OnnxModel.py``` file. Moving it.

File: pip-dep/requirements.txt (hunk 1 above)
Comment (LaRiffle) [reply pip-dep/requirements.txt:15]: v0.4.10 of syft proto has been released with ONNX support

Comment (gmuraru) [reply pip-dep/requirements.txt]: Done -- waiting for tests

File: test/crypten/test_context.py (hunk 10 above)
Comment (gmuraru) [reply test/crypten/test_context.py:236]: Done
